modules/040-node-manager/images/bashible-apiserver/pkg/template/context.go

```diff
@@ -422,3 +454,188 @@ func (c *BashibleContext) saveChecksum(name, newChecksum string) {
 
 	c.checksums[name] = newChecksum
 }
+
+// newBashibleInformerFactory creates informer factory for particular namespace and label selector.
+// Bashible apiserver is expected to use single namespace and only related resources.
+func newBashibleInformerFactory(kubeClient client.Client, resync time.Duration, namespace, labelSelector string) informers.SharedInformerFactory {
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		kubeClient,
+		resync,
+		informers.WithNamespace(namespace),
+		informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+			opts.LabelSelector = labelSelector
+		}),
+	)
+
+	return factory
+}
+
+func newNodeUserInformerFactory(kubeClient client.Client, resync time.Duration) dynamicinformer.DynamicSharedInformerFactory {
+	factory := dynamicinformer.NewDynamicSharedInformerFactory(
+		kubeClient.Dynamic(),
+		resync,
+	)
+
+	return factory
+}
+
+func (c *BashibleContext) subscribeOnNodeUserCRD(ctx context.Context, ngConfigFactory dynamicinformer.DynamicSharedInformerFactory) {
+	if ngConfigFactory == nil {
+		return
+	}
+
+	go c.runBufferedEmitter()
+	go c.runNodeUserCRDQueue(ctx)
+
+	// Launch the informer
+	ginformer := ngConfigFactory.ForResource(schema.GroupVersionResource{
+		Group:    "deckhouse.io",
+		Version:  "v1",
+		Resource: nodeUserCRDName,
+	})
+
+	informer := ginformer.Informer()
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.nodeUsersQueue <- usersQueueAction{
+				action:    "add",
+				newObject: obj.(*unstructured.Unstructured),
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.nodeUsersQueue <- usersQueueAction{
+				action:    "update",
+				newObject: newObj.(*unstructured.Unstructured),
+				oldObject: oldObj.(*unstructured.Unstructured),
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.nodeUsersQueue <- usersQueueAction{
+				action:    "delete",
+				oldObject: obj.(*unstructured.Unstructured),
+			}
+		},
+	})
+
+	go informer.Run(ctx.Done())
+
+	// Wait for the first sync of the informer cache, should not take long
+	if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+		klog.Fatalf("unable to sync caches: %v", ctx.Err())
+	}
+}
+
+func (c *BashibleContext) runBufferedEmitter() chan struct{} {
```

Comment:
Maybe we can move this emitter to some common code to not copy the code between different packages

Reply:
its a good point, i refactor this code.

Reply:
good point, i refactor them